finality-aleph/src/data_io.rs

```diff
@@ -13,73 +13,256 @@ use futures::{
+    // an ascending sequence of blocks b_0, b_1, ..., b_n satisfying the following conditions:
+    //     1) hash(b_i) = h_i for i = 0, 1, ..., n,
+    //     2) parent(b_{i+1}) = b_i for i = 0, 1, ..., (n-1),
+    //     3) height(b_n) = num,
+    //     4) The parent of b_0 has been finalized (prior to creating this AlephData).
+    // Such an AlephData object should be thought of as a proposal for block b_n to be finalized.
+    pub branch: Vec<B::Hash>,
+    pub number: NumberFor<B>,
+}
+
+// Need to be implemented manually, as deriving does not work (`BlockT` is not `Hash`).
+impl<B: BlockT> Hash for AlephProposal<B> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.branch.hash(state);
+        self.number.hash(state);
+    }
+}
+
+// Clippy does not allow deriving PartialEq when implementing Hash manually
+impl<B: BlockT> PartialEq for AlephProposal<B> {
+    fn eq(&self, other: &Self) -> bool {
+        self.number.eq(&other.number) && self.branch.eq(&other.branch)
+    }
+}
+
+impl<B: BlockT> Eq for AlephProposal<B> {}
+
+impl<B: BlockT> AlephProposal<B> {
+    pub(crate) fn new(branch: Vec<B::Hash>, block_number: NumberFor<B>) -> Self {
+        AlephProposal {
+            branch,
+            number: block_number,
+        }
+    }
+
+    pub(crate) fn is_correct(&self) -> bool {
+        if self.branch.len() > MAX_DATA_BRANCH_LEN as usize {
+            return false;
+        }
+        if self.branch.is_empty() {
+            return false;
+        }
+        if self.number < <NumberFor<B>>::saturated_from(self.branch.len()) {
+            // Honest nodes always propose a block strictly above the most recently finalized one. Also
+            // the bottom block of the branch they choose as one above the most recently finalized one.
+            return false;
+        }
+        true
+    }
+
+    pub(crate) fn top_block(&self) -> BlockHashNum<B> {
+        // Assumes that the data is correct
+        BlockHashNum::<B>::new(
+            *self
+                .branch
+                .last()
+                .expect("cannot be empty for correct data"),
+            self.number,
+        )
+    }
+    pub(crate) fn bottom_block(&self) -> BlockHashNum<B> {
+        // Assumes that the data is correct
+        BlockHashNum::<B>::new(
+            *self
+                .branch
+                .first()
+                .expect("cannot be empty for correct data"),
+            self.number,
+        )
+    }
+
+    pub(crate) fn number_below_branch(&self) -> NumberFor<B> {
+        // Assumes that data is correct
+        self.number - <NumberFor<B>>::saturated_from(self.branch.len())
+    }
+
+    pub(crate) fn number_bottom_block(&self) -> NumberFor<B> {
+        // Assumes that data is correct
+        self.number - <NumberFor<B>>::saturated_from(self.branch.len() - 1)
+    }
+
+    pub(crate) fn number_top_block(&self) -> NumberFor<B> {
+        // Assumes that data is correct
```

Comment:
Changed `is_correct` to `within_bounds` and added a lot of comments. 

I don't feel like prefixing with `expected` clarifies anything here. This data can be malicious and thus meaningless, it is not clear why this given output should be `expected`. `claimed` would be perhaps a better prefix, but still -- I don't think it would help anyone to understand the code.